Sources/SourceKitLSP/Swift/SemanticTokens.swift

```diff
@@ -0,0 +1,352 @@
+      return (.class, [.declaration])
+    case values.ref_class:
+      return (.class, [])
+    case values.decl_struct:
+      return (.struct, [.declaration])
+    case values.ref_struct:
+      return (.struct, [])
+    case values.decl_enum:
+      return (.enum, [.declaration])
+    case values.ref_enum:
+      return (.enum, [])
+    case values.decl_protocol:
+      return (.interface, [.declaration])
+    case values.ref_protocol:
+      return (.interface, [])
+    case values.decl_associatedtype,
+         values.decl_typealias,
+         values.decl_generic_type_param:
+      return (.typeParameter, [.declaration])
+    case values.ref_associatedtype,
+         values.ref_typealias,
+         values.ref_generic_type_param:
+      return (.typeParameter, [])
+    case values.decl_function_constructor,
+         values.decl_function_subscript,
+         values.decl_function_free:
+      return (.function, [.declaration])
+    case values.decl_function_method_static,
+         values.decl_function_method_instance,
+         values.decl_function_method_class:
+      return (.method, [.declaration])
+    case values.ref_function_constructor,
+         values.ref_function_destructor,
+         values.ref_function_free,
+         values.ref_function_subscript:
+      return (.function, [])
+    case values.ref_function_method_static,
+         values.ref_function_method_instance,
+         values.ref_function_method_class:
+      return (.method, [])
+    case values.decl_function_operator_prefix,
+         values.decl_function_operator_postfix,
+         values.decl_function_operator_infix:
+      return (.operator, [.declaration])
+    case values.ref_function_operator_prefix,
+         values.ref_function_operator_postfix,
+         values.ref_function_operator_infix:
+      return (.operator, [])
+    case values.decl_var_static,
+         values.decl_var_class,
+         values.decl_var_instance:
+      return (.property, [.declaration])
+    case values.ref_var_static,
+         values.ref_var_class,
+         values.ref_var_instance:
+      return (.property, [])
+    case values.decl_var_local,
+         values.decl_var_global:
+      return (.variable, [.declaration])
+    // We ignore `value.decl_var_parameter`s for now.
+    // SourceKit seems to use these to refer to parameter labels,
+    // therefore we don't use .parameter here (which LSP client like
+    // VSCode seem to interpret as variable identifiers, however
+    // causing a 'wrong highlighting' e.g. of `x` in `f(x y: Int) {}`)
+    case values.ref_var_local,
+         values.ref_var_global:
+      return (.variable, [])
+    case values.syntaxtype_comment,
+         values.syntaxtype_doccomment:
+      return (.comment, [])
+    case values.syntaxtype_type_identifier:
+      return (.type, [])
+    case values.syntaxtype_number:
+      return (.number, [])
+    case values.syntaxtype_string:
+      return (.string, [])
+    default:
+      return nil
```

Comment:
What do you think about adding some low-level logging here if we encounter a kind that’s not handled, e.g. on the warning level? That way we would have a way of noticing if new kinds are added to SourceKit that we don’t handle here yet. In that case you’d need to add a case for `value.decl_var_parameter` that returns `nil` without logging.

---

I think you’re also missing a few syntax kinds here. For example, there’s no `syntaxtype_identifier`. You could probably catch most of them by looking through [UIDs.py](https://github.com/apple/swift/blob/main/utils/gyb_sourcekit_support/UIDs.py) again and by enabling the logging.

Reply:
Is there a way to log these uids with their string representation?

`syntaxtype_identifier` is one I deliberately ignored for now since it's kind of ambiguous (is it a variable? a function? a method? a parameter?) and makes the syntax highlighting actually less specific than VSCode's default highlighter.

Reply:
You should be able to use `uid_get_string_ptr` like here:
https://github.com/apple/sourcekit-lsp/blob/13d3d93d8b706ff93bd88644aa1a4b6a2c054a81/Sources/SourceKitLSP/Swift/SemanticRefactorCommand.swift#L75